src/irmin-pack/unix/pack_store.ml

```diff
@@ -420,22 +420,28 @@ struct
     Lwt.try_bind (fun () -> f (cast t)) on_success on_fail
 
   let unsafe_append ~ensure_unique ~overcommit t hash v =
+    let kind = Val.kind v in
+    let use_index =
+      Irmin_pack.Indexing_strategy.is_minimal t.indexing_strategy
+      && kind = Commit_v2
+    in
     let unguarded_append () =
-      [%log.debug "[pack] append %a" pp_hash hash];
+      let check_in_index hash =
+        (* TODO: Why don't we promote the key here? *)
+        match Index.find (Fm.index t.fm) hash with
+        | None ->
+            Stats.incr_appended_hashes ();
+            None
+        | Some (offset, _, _) ->
+            Stats.incr_appended_offsets ();
+            Some offset
+      in
       let offset_of_key k =
         match Pack_key.inspect k with
         | Direct { offset; _ } ->
             Stats.incr_appended_offsets ();
             Some offset
-        | Indexed hash -> (
-            (* TODO: Why don't we promote the key here? *)
-            match Index.find (Fm.index t.fm) hash with
-            | None ->
-                Stats.incr_appended_hashes ();
-                None
-            | Some (offset, _, _) ->
-                Stats.incr_appended_offsets ();
-                Some offset)
+        | Indexed hash -> if use_index then check_in_index hash else None
```

Comment:
I'm not sure its a problem, but can't we have cases where the store switches from always to minimal and reuses "old" objects that are indexed, while constructing "new" ones?

Reply:
Do you think that could be an issue with the GC? As we can't track offsets properly? But in that case iter will look in the index, right?

Reply:
Of so this indeed an issue: https://github.com/mirage/irmin/blob/main/src/irmin-pack/unix/gc.ml#L254

Reply:
I've reverted this change

Reply:
The situation I had in mind is not on the gc, but on normal operations, but I now think its not a problem (when a `always` store switches to `minimal`, it appends a new objects they all use the new strategy, so it should be ok).